### ISSUE TITLE
Error: Cannot find module 'eslint-plugin-react'

### ISSUE THREAD

**Issue by studiorgb** [eslint/eslint#3782]
Fresh global installation, first test and error :( Any ideas?

```
pawelgrzybek$ eslint js/script1.js
module.js:338
    throw err;
    ^

Error: Cannot find module 'eslint-plugin-react'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:286:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at /usr/local/lib/node_modules/eslint/lib/cli-engine.js:116:26
    at Array.forEach (native)
    at loadPlugins (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:107:21)
    at processText (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:194:5)
    at processFile (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:254:18)
    at executeOnFile (/usr/local/lib/node_modules/eslint/lib/cli-engine.js:540:23)
pawelgrzybek$
```

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by studiorgb** [comment]
Sorry. That solved the problem.
```npm install -g eslint-plugin-react```

**Comment by snowcxt** [comment]
Just update to 5.0.0. I got the error Cannot find module 'eslint-plugin-react' and  `npm install -g eslint-plugin-react` can't resolve this issue.

**Comment by platinumazure** [comment]
ESLint isn't at 5.0.0 yet. What did you actually update?

**Comment by mysticatea** [comment]
Maybe it's about Atom's `linter-eslint`.
If so, I'd like to recommend you to open an issue in https://github.com/AtomLinter/linter-eslint/issues

**Comment by IanVS** [comment]
The issue has been reported at: https://github.com/AtomLinter/linter-eslint/issues/259

**Comment by mysticatea** [comment]
Oh, thank you!

**Comment by snowcxt** [comment]
Yes, thank you so much!

**Comment by adjavaherian** [comment]
@pawelgrzybek seems to have nailed it. `npm install -g eslint-plugin-react`

**Comment by DelvarWorld** [comment]
I ran into this error too and `rm -rf node_modules`, `npm cache clean` and `npm cache clear` did nothing. For me the issue was some combination of nvm and global eslint. For me, I had to `npm uninstall -g eslint`, then install it for the specific project `npm install eslint`, then close my shell and open a new session, because in my old shell when I ran `eslint` it kept trying to run my old global version in the nvm node_modules directory.

tl;dr I don't know where I am or what's happening

**Comment by IanVS** [comment]
The basic thing to keep in mind is that the plugins need to be installed wherever you are running eslint.  If you are typing `eslint` into your terminal, you are running the globally installed eslint (unless of course you've set up an alias or changed your PATH variable, but that's rare).

Typically we suggest that you install eslint and any plugins locally (`npm install --save-dev eslint eslint-plugin-react` without the `-g` flag).  Then, set up an npm script as something like this:

```json
"scripts": {
    "lint": "eslint ."
}
```

Be sure to use whatever directory and options apply to you.  Then, when you want to lint your files, use `npm run lint`, and npm will use the local version and won't try to run a global eslint.

The other way to run the local eslint is a little more clunky, but can be done with:

```
$ node_modules/.bin/eslint .
```

Does this help at all?

**Comment by mrbinky3000** [comment]
FWIW.  I quickly set up a local dev environment and ran into this error.   The error said it could not find `eslint-plugin-react` and it was installed.  However, a few minutes later, I realized that I never installed `eslint`.  The error message was a little misleading.   Saying it can't find a plugin implies that you actually installed the main module itself.  For me, this was not the case.

**Comment by IanVS** [comment]
You must have had eslint installed somewhere, maybe globally?  How did you run eslint to get the error if it wasn't installed?

**Comment by eslintbot** [comment]
Hi @BYK, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by mrbinky3000** [comment]
@BYK.  You are correct, I just discovered that eslint was installed globally, but not locally, so the reason it couldn't find eslint-plugin-react was because the global version of eslint couldn't fine the locally installed version, which is correct behavior.

**Comment by jeffwesson** [comment]
Thanks @IanVS and @DelvarWorld 
It was a combination of: 
```bash
rm -rf node_modules && npm uninstall -g eslint eslint-plugin-react
``` 
and 
```bash
npm install --save-dev eslint eslint-plugin-react && npm install
```
Added the trailing npm install to compensate for rm'ing node_modules entirely, just in case anyone copy/pastes this without looking :wink:

**Comment by dgilliam** [comment]
thanks @jeffwesson !

**Comment by liyaodong** [comment]
solved by ⬇️
```
$ cd ~/.atom/packages/linter-eslint
$ npm i eslint-plugin-react
```

**Comment by themoritz** [comment]
@liyaodong's solution also solved the issue for me. Thanks!

**Comment by IanVS** [comment]
^ While this will work, it is not the best way to address the issue.  This installs `eslint-plugin-react` into `linter-eslint` directly.  The reason it's solving it for you is that you don't have ESLint installed in your own project and you don't have ESLint installed globally, so `linter-eslint` is falling back to the the version of ESLint that it bundles as a backup.  There have been discussions that we may stop bundling it in the future, because of the confusion it causes, not knowing what ESLint you're actually using.  

The best thing to do is to install eslint and any necessary plugins into your project with `npm install --save-dev eslint eslint-plugin-react`.  This will ensure that everyone contributing to your project is using the right version of ESLint.

**Comment by themoritz** [comment]
You're right @IanVS, that also immediately solved it. Thanks!

**Comment by jpalala** [comment]
Interesting documentation showed up upon error:

`> eslint --config ./browser.eslintrc --ignore-path browser.eslintignore ./public/static/app/js`

```
Oops! Something went wrong! :(

ESLint couldn't find the plugin "eslint-plugin-react". This can happen for a couple different reasons:

1. If ESLint is installed globally, then make sure eslint-plugin-react is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.

2. If ESLint is installed locally, then it's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm i eslint-plugin-react@latest --save-dev
```

**Comment by jpalala** [comment]
![eslint-interesting](https://cloud.githubusercontent.com/assets/622659/15771813/c60b8138-299f-11e6-81e7-5962e9430d6f.png)

**Comment by hafuta** [comment]
For those of you who still have difficulty finding the cause for this error when using the eslint plugin for VSCode, I had the same issue, eslint and eslint-plugin-react were installed locally, I verified that I don't have the eslint and the plugin installed globally but the problem still happened.
Running eslint from local installation (./node_modules/.bin/eslint) was working fine.

I've managed to solve the issue only after deleting the node_modules directory from my home directory, which is strange, this folder shouldn't be in use.

**Comment by stoplion** [comment]
Having same problem the above solutions did not fix

**Comment by kevinSuttle** [comment]
> I've managed to solve the issue only after deleting the node_modules directory from my home directory 

This fixed Atom for me.

**Comment by coldacid** [comment]
Why does eslint need a plugin to run when it's not configured to use it (or any other plugins which depend on it)?

**Comment by platinumazure** [comment]
@coldacid The short answer is that ESLint probably *is* configured to use that plugin, and you just might not know it due to the use of a shareable config or plugin (which may itself use another shareable config or plugin, etc.). You can use `eslint --print-config someFile.js` to get your full, effective configuration for that file. If you're experiencing a specific issue, though, please open a new issue.

**Comment by ummahusla** [comment]
@liyaodong Nice one, thanks!

**Comment by a7madgamal** [comment]
same error, npm install -g eslint-plugin-react fixed it.
However, I think the error should be more friendly. a nice informative error popup to run "npm install -g eslint-plugin-react" is much more easier than opening the devtools, google, reading this thread and finding the comment :)

**Comment by matthew-dean** [comment]
In my case, I'm not able to install the plugin globally.

Use case:

# I'm depending on a private module that essentially has building / linting scripts included.
# Said scripts in `node_modules/private-repo` run ESLint, but grab from `../../.eslintrc` (project root) for configuration.
# Project root `.eslintrc` has a plugin "eslint-plugin-require-jsdoc-except". (The name of the plugin is irrelevant. It's the same error as this thread.) So, `npm start` runs `./node_modules/private-repo/package-scripts.js` which runs `eslint --config /[projectroot]/.eslintrc`. At that point, ESLint is unable to resolve the path of the plugin.
# Interestingly, a DIFFERENT ESLint plugin that is in the `package.json` for `node_modules/private-repo` is able to be resolved by ESLint. But the one from the project root `package.json` is not.

**Comment by platinumazure** [comment]
@matthew-dean I'm not sure I follow. In order for the private repo to be able to lint itself, standalone, it would need to have the dependent package in its own package.json. Why not just specify the ESLint plugin as a dependency of both projects, so that either project could lint itself?

Please stop by our [Gitter chat](https://gitter.im/eslint/eslint) if you need more help.

**Comment by matthew-dean** [comment]
@platinumazure 

1. Project A specifies Build module B
2. Build module B imports ESLint config from containing project (Project A) as part of build scripts (to enable extensibility per project).
3. Project A `.eslintrc` adds an ESLint plugin, which is installed as a dependency of Project A.
4. ESLint cannot find plugin specified in Project A's `.eslintrc` file, even though ESLint is run from the Project A folder. ESLint is installed locally in both Project A's dependencies and Build module B's dependencies.
